src/scales/scale.time.js

```diff
@@ -272,23 +272,51 @@ function determineStepSize(min, max, unit, capacity) {
 	return factor;
 }
 
+function determineMajorUnit(unit) {
+	for (var i = UNITS.indexOf(unit) + 1, ilen = UNITS.length; i < ilen; ++i) {
+		if (INTERVALS[UNITS[i]].common) {
+			return UNITS[i];
+		}
+	}
+}
+
 /**
  * Figures out what unit results in an appropriate number of auto-generated ticks
  */
-function determineUnitForAutoTicks(minUnit, min, max, capacity) {
+function determineUnitsForAutoTicks(scale) {
+	var timeOpts = scale.options.time;
+	var minor = timeOpts.unit;
+	var max = scale.max;
+	var range = max - scale.min;
 	var ilen = UNITS.length;
-	var i, interval, factor;
-
-	for (i = UNITS.indexOf(minUnit); i < ilen - 1; ++i) {
-		interval = INTERVALS[UNITS[i]];
-		factor = interval.steps ? interval.steps[interval.steps.length - 1] : MAX_INTEGER;
-
-		if (interval.common && Math.ceil((max - min) / (factor * interval.size)) <= capacity) {
-			return UNITS[i];
+	var i, major, interval, steps, factor, capacity;
+
+	if (minor) {
+		major = determineMajorUnit(minor);
```

Comment:
Another issue I discovered is that we don't always want the major unit to be just one unit larger than the minor unit. E.g. I was plotting a chart with 8000 days and then it made more sense to just plot years as labels by making the majorUnit by year. I was attempting to fix that issue in https://github.com/chartjs/Chart.js/pull/6274 by ensuring that the major unit fit on the scale by calling `determineUnitsForAutoTicks`. That gets to be a bit difficult with the way that this PR is changing the method. Any ideas on how we might address both that issue as well as the ones you're addressing here?

Here's the change I was making https://github.com/chartjs/Chart.js/pull/6274/files#diff-4cc62430ebd939b9632b5bbc890c79b4R580. I'd appreciate any review you can give on that PR as well. Thanks!

Reply:
Could add a option for major unit, like in `ticks.marjor` or `time.marjorUnit`? That would allow any configuration desired.

Reply:
I did a quick review of #6274. Because we don't know which unit is going to be a major unit [at the first `determineUnitForAutoTicks` call](https://github.com/chartjs/Chart.js/blob/889ae8727ea74bed2b62f6008672898f31f52ffe/src/scales/scale.time.js#L296), one possible solution is to pick the minimum capacity among the minor and all other larger units (e.g. if the minor unit is `'day'`, check for capacities of `'day'` using the minor tick font and `'month'`/`'year'` using the major tick font, then use the minimum one). I think this will generate a better result than always using the `'millisecond'` format.

Anyway, it seems that #6274 should be merged first, then I will make required changes here.